### PR TITLE
feat: Passwordless SMS code authentication

### PR DESCRIPTION
**Pull request by splaunov**
Implements new "SMS" login strategy. This strategy requires two login flow submissions:

- the first one is to send a phone number to the server
- and the second - to send the received SMS code for validation

## Related issue(s)

#1570

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by splaunov** [comment]
@alexey-reshetnik Hi! I have merged code from your PR #1941 to take advantage of sending authentication SMS codes through courier.

**Comment by aeneasr** [comment]
Thank you for continuing the work on this! Tag me for a review once you want it :)

**Comment by alexey-reshetnik** [comment]
Thank you very much for picking up the development. I fell out from work for couple of weeks due to family emergency

**Comment by aeneasr** [comment]
Thank you, to be honest I have a bit of a hard time reviewing the PR because it is so large. Could you maybe explain what the PR solves in particular? Adding documentation could also be very helpful to understand e.g. how to set it up!

File: persistence/sql/persister_test.go [hunk 1]
```diff
@@ -153,49 +144,53 @@ func TestPersister(t *testing.T) {
...
-			t.Run("contract=recovery.TestFlowPersister", func(t *testing.T) {
-				pop.SetLogger(pl(t))
-				recovery.TestFlowPersister(ctx, conf, p)(t)
-			})
-			t.Run("contract=link.TestPersister", func(t *testing.T) {
-				pop.SetLogger(pl(t))
-				link.TestPersister(ctx, conf, p)(t)
-			})
-			t.Run("contract=continuity.TestPersister", func(t *testing.T) {
+			//t.Run("contract=identity.TestPool", func(t *testing.T) {
```
Comment (aeneasr) [inline persistence/sql/persister_test.go:147]: Please activate the tests again

File: schema/extension.go [hunk 2]
```diff
@@ -20,6 +20,9 @@ type (
 			Password struct {
 				Identifier bool `json:"identifier"`
 			} `json:"password"`
+			Sms struct {
```
Comment (aeneasr) [inline schema/extension.go:23]: Convention says that abbreviations are upper case in golang:

```suggestion
			SMS struct {
```

**Comment by aeneasr** [comment]
Another important topic is probably wether this "code" strategy can be used for email also? So https://github.com/ory/kratos/issues/1451?

**Comment by aeneasr** [comment]
While the PR is being worked on I will mark it as a draft. That declutters our review backlog :)

Once you're done with your changes and would like someone to review them, mark the PR as ready and request a review from one of the maintainers.

Thank you!

**Comment by aeneasr** [comment]
One of the PRs by @alexey-reshetnik was now merged :) The other one (courier) is also looking very solid!

File: persistence/sql/persister_test.go (hunk 1 above)
Comment (splaunov) [reply persistence/sql/persister_test.go]: Done

File: schema/extension.go (hunk 2 above)
Comment (splaunov) [reply schema/extension.go]: Renamed `SMS` to `Code`

**Comment by splaunov** [comment]
> Another important topic is probably wether this "code" strategy can be used for email also? So #1451?

Yes, great idea. Have renamed the strategy from `sms` to `code` to be able to use it with emails also.

**Comment by yingy77** [comment]
When logging in with the same phone number multiple times, there are duplicate records in the identities table, which seems to be wrong. I modified the selfservice/strategy/sms/login.go file and it seems to have solved the problem.
[fix-sms_login-multi_identities.zip](https://github.com/ory/kratos/files/7931943/fix-sms_login-multi_identities.zip)
.

**Comment by splaunov** [comment]
@yingy77 thanks for reporting the issue!
Should have been fixed for now.

**Comment by splaunov** [comment]
Have added verification flow to the `code` strategy.

**Comment by splaunov** [comment]
I have a requirement to add fixed-code logins for testing purposes and to ease apple review process for iOS apps.
My suggestion is to add list of fake phone numbers and a fixed code to the `code` strategy config.
So, testers and reviewers could login with the known phone number and the fixed code.

@aeneasr what do you think about this?
This is highly essential for apps, where phone login is the only auth method.

**Comment by aeneasr** [comment]
Hey there, I am currently working on passwordless auth using webauthn. I suggest that once it is merged we copy that pattern for SMS auth: https://github.com/ory/kratos/pull/2260

**Comment by splaunov** [comment]
> Hey there, I am currently working on passwordless auth using webauthn. I suggest that once it is merged we copy that pattern for SMS auth: #2260

@aeneasr can you summarise what are the key points of the pattern?
Should I update this PR according to what is done in #2260 or what are your plans regarding SMS auth?

**Comment by langovoi** [comment]
@splaunov Any progress here? Maybe you needs help?

**Comment by splaunov** [comment]
Really need response from @aeneasr to my previous post. Not sure how to move forward here.

**Comment by aeneasr** [comment]
> > Hey there, I am currently working on passwordless auth using webauthn. I suggest that once it is merged we copy that pattern for SMS auth: #2260
> 
> @aeneasr can you summarise what are the key points of the pattern? Should I update this PR according to what is done in #2260 or what are your plans regarding SMS auth?

TO this?

**Comment by splaunov** [comment]
Yes. Or maybe just let's review what is done in this PR and then decide about next steps.

**Comment by aeneasr** [comment]
Ok, I'm back from vacation and have time to review this :)

**Comment by sidharthramesh** [comment]
This would be a great feature to have!
Most of our users don’t have emails and rely on mobile phones as their primary means of identification.

**Comment by shasha-zhang** [comment]
Any updates for this? I agree this is a great feature. A lot of people like to login with code sent through SMS.

**Comment by ankit-shopflo** [comment]
Any update on this feature? We are planning to move from firebase, the only blocker is this SMS feature

**Comment by vinckr** [comment]
> Any update on this feature? We are planning to move from firebase, the only blocker is this SMS feature

No update, contributions welcome.

**Comment by nicolasburtey** [comment]
> Any update on this feature? We are planning to move from firebase, the only blocker is this SMS feature

same for us I guess, waiting for this to be merged to been able to use kratos

**Comment by mohdalali** [comment]
I am also waiting for this to be able to use Kratos. Would like to see this soon!

**Comment by adgang** [comment]
May I please know why this ubiquitous use case is still not in kratos. Even after we have a PR? Is there anything we can do to help this feature along?

**Comment by nair-sreerag** [comment]
@aeneasr Can u pls look into this and merge it with the main branch?? The absence of this (much required) feature is the only thing that is keeping us from migrating to Ory .

**Comment by CNLHC** [comment]
It seems that the Passwordless SMS  code flow is quite similar to  #3378, especially with the help of the `courier` layer. Both the SMS and the email need to send a one-time code to the user, the only difference is the delivery method.

Since #3378 is merged (and works well in our deployment), I am a little curious about why this PR needs to change 5000 LOC to implement the Passwordless SMS code authentication.

**Comment by aeneasr** [comment]
This feature is now available in Ory Kratos! :) Thank you for your contribution

**Comment by aeneasr** [comment]
Sorry, I thought this was email code login, which is supported (sms passwordless login is afaik not yet supported).

**Comment by wgorczyca** [comment]
any update about this feature?

**Comment by aeneasr** [comment]
This should now be possible on master by using an sms channel and code with passwordless!

I‘m closing this PR as it is obsolete

**Pull request closed without merging**